src/alternatives.js

```diff
@@ -0,0 +1,144 @@
+import has from 'lodash/has';
+import inherits from './util/inherits';
+import MixedSchema from './mixed';
+import { alternatives as locale } from './locale.js';
+import isSchema from './util/isSchema';
+import typeOf from 'type-name';
+import { SynchronousPromise } from 'synchronous-promise';
+import createValidation from './util/createValidation';
+import runValidations from './util/runValidations';
+
+let promise = sync => sync ? SynchronousPromise : Promise;
+const invalidMarker = {};
+
+
+export default function AlternativesSchema() {
+    if (!(this instanceof AlternativesSchema))
+        return new AlternativesSchema()
+
+    MixedSchema.call(this, { type: 'alternatives' })
+}
+const defaultShouldUseForCasting = () => true;
+
+inherits(AlternativesSchema, MixedSchema, {
+    _typeCheck(v){
+        if(this._oneOfTypeOptions){
+            const {
+                schemas
+            } = this._oneOfTypeOptions;
+            return !!schemas.find(it => it.isType(v));
+        }
+        return false
+    },
+
+    _validate(_value, options = {}) {
+        let value = _value;
+        let originalValue = options.originalValue != null ?
+            options.originalValue : _value
+
+        let isStrict = this._option('strict', options)
+        let endEarly = this._option('abortEarly', options)
+
+        let sync = options.sync
+        let path = options.path
+        let label = this._label
+
+        if (!isStrict) {
+            value = this._cast(value, { assert: false, ...options })
+        }
+        // value is cast, we can check if it meets type requirements
+        let validationParams = { value, path, schema: this, options, label, originalValue, sync }
+
+        const validations = [];
+
+        if(this._oneOfType){
+            validations.push(this._oneOfType(validationParams));
+        }
+
+        return runValidations({
+            validations: validations, endEarly, value, path, sync
+        });
+    },
+
+    _cast(rawValue, options = {}) {
+        let value = rawValue;
+        if (this._oneOfTypeOptions) {
+            const {
+                schemas,
+                shouldUseForCasting = defaultShouldUseForCasting
+            } = this._oneOfTypeOptions;
+            for (let schema of schemas) {
+                // It's difficult to decide what type we want to use for casting
+                // the value. Ultimately, it's up to the developer so allow the developer
+                // to decide using `shouldUseForCasting` (defaults to for all types).
+                if(shouldUseForCasting(schema, value, options)){
```

Comment:
I'm not sure I understand this toggle, ideally the schema you use cast should be the same one used to validate, you don't want to cast a value to a Date and then validate it as a number for instance

Reply:
Ideally yes, but there is loss of information. See this example:
```
      inst.validateSync('10',{strict:true}).should.equal('10');
      inst.cast('10').should.equal(10)
```
The first is handled by `string()` and the latter by `number()`. So I created an escape hatch for developers who want to influence this. 
Better would be to validate before cast and throw an Error if a promise is used to validate and inform that validate should be used for casting with promise validators. Or to have cast return a promise as well, but that would be breaking and a bad idea.